File: fairies.java
import java.util.*;
public class fairies {
	public static void main(String[] args){
		Scanner in = new Scanner(System.in);
		int t = in.nextInt();
		
		/*
		 * These numbers are called Catalan numbers. Since we only want to know the minimum number of
		 * groups that the faries can form, we really just need to know if the sum of all the Catalan 
		 * numbers in a range is even or odd. If it's even, we need to form an extra group because we 
		 * can't put a fairy at the tip of the group if every fairy is part of a pair. So, if the sum 
		 * is even we must form two groups. Otherwise, we can form one. Since we only need to know if
		 * the sum of the Catalan numbers in the given range is even or odd, we really only need to
		 * know how many odd Catalan numbers are part of the range. If we look at the first several
		 * Catalan numbers, a pattern becomes apparent: Catalan numbers are odd if their index is one
		 * less than a power of two. So, the problem becomes figuring out how many numbers that are 
		 * one less than a power of two are in the given range.
		 */
		
		//Here we're precomputing all of the powers of two less than 10e18, and adding them to a list.
		ArrayList<Long> twos = new ArrayList<Long>();
		long last = 1;
		for(int i = 1; i < 63; i++){
			//The -1 is because we're actually looking for numbers one less than a power of two.
			twos.add(last - 1);
			last*=2;
		}
		for(int z = 0; z < t; z ++){
			long a = in.nextLong();
			long b = in.nextLong(); 
			
			//Now let's go through our list of valid indices and see how many are in our range. 
			int ctr = 0;
			for(int i = 0; i < twos.size(); i ++){
				if(twos.get(i) >= a && twos.get(i) <= b){
					ctr++;
				}
			}
			
			/*
			 * If the number of odd Catalan numbers is even, we want to say that we formed two 
			 * groups. If it's odd, we want to say that we formed one.
			 */
			ctr++;
			System.out.println((ctr % 2) + 1);
		}
	}
}

File: fairies.c
#include <stdio.h>
#include <stdlib.h>
/* There are three cases:
   0 faries fly in 0 groups
   an odd number of fairies fly in 1 group
   an even numbe of fairies flys in 2 groups (1 fairy seperate and the rest together)
   
   0 fairies never happens (input 0 0 involves one fairy) so we only care about if 
   the number of fairies that can live in the trees is even or odd.
   
   Now, to determine if there the number of fairies in trees 0-N (inclusive) is odd or even
   Because of how ridiciously large the input is, we know there is a pattern or this would be
   unsolvable.  We can write a program to output the catalan numbers in order and see what
   the pattern is. We can also try and logic out how often you add an odd number of odd numbers
   but this is much harder unless mathmatics is your thing.

   We find from this the C(0), C(1), C(3), C(7), and C(15) are all odd which are all one less than a
   power of two.  This means that the sum of all catalans from 0-N switches between even or odd at
   every time the catalan number for the final tree is one less than a power of 2 (probably, but only
   question authors need to write formal proofs!).

   To decide if the sum of all catalans from 0-N is even or odd, we need to know the last time it
   switched, aka the largest power of 2 less than N.  This can be found by figuring out
   what the left most one bit in N+1 is since each bit in N represents a power of two less than
   or equal to it.  If that bit is in an odd posistion the sum is odd otherwise it is even.

   Once we know if, for our input values a and b, 0-a and 0-b are even or odd we can look at
   the range a to b as 0 to b minus 0 to (a-1).  Odd minus even or even minus odd is odd,
   all others are even.  We model this as leftBit(a)+leftBit(b-1) % 2. Now when we consider we
   add one before our check, the answer is ACTUALLY found as leftbit(a+1)+leftBit(b) % 2.
*/

int leftBit(unsigned long long a);

int main() {
    int numForests;
    int caseNum = 0;
    scanf("%d", &numForests);
    
    //iterate through all case numbers
    while (caseNum ++ < numForests)
    {
	//unsigned long long holds up to 2^64
	unsigned long long a, b;
	scanf("%llu %llu", &a, &b);

	//check to make sure input follows bounds
	if (a > b) 
	{
	    printf("error! ");
	}

	//parity is if a number is odd
	int parity = (leftBit(b+1) + leftBit(a)) %2;

	//print one if odd, two if even
	printf("%d\n", parity == 1 ? 1 : 2);
    }
}

//gets the left most bit
int leftBit(unsigned long long a)
{
    int result = 0;
    //while there is a bit more left than what we are at
    while (a > 0)
    {
        //increment our left bit counter and move the number right one bit
	result++;
	a /= 2;
    }

    return result;
}
